lgc/tool/lgc/lgc.cpp

```diff
@@ -0,0 +1,259 @@
+//
+// @param data : Input data to check
+static bool isElfBinary(StringRef data) {
+  bool isElfBin = false;
+  if (data.size() >= sizeof(ELF::Elf64_Ehdr)) {
+    auto pHeader = reinterpret_cast<const ELF::Elf64_Ehdr *>(data.data());
+    isElfBin = pHeader->checkMagic();
+  }
+  return isElfBin;
+}
+
+// =====================================================================================================================
+// Checks whether the output data is actually ISA assembler text
+//
+// @param data : Input data to check
+static bool isIsaText(StringRef data) {
+  // This is called by the lgc standalone tool to help distinguish between its three output types of ELF binary,
+  // LLVM IR assembler and ISA assembler. Here we use the fact that ISA assembler is the only one that starts
+  // with a tab character.
+  return data.startswith("\t");
+}
+// =====================================================================================================================
+// Main code of LGC standalone tool
+//
+// @param argc : Count of command-line arguments
+// @param argv : Command-line arguments
+int main(int argc, char **argv) {
+  const char *progName = sys::path::filename(argv[0]).data();
+  LLVMContext context;
+  LgcContext::initialize();
+  cl::ParseCommandLineOptions(argc, argv, "lgc: command-line tool for LGC, the LLPC middle-end compiler\n"
+      "\n"
+      "The lgc tool parses one or more modules of LLVM IR assembler from the input\n"
+      "file(s) and compiles each one using the LGC interface, into AMDGPU ELF or\n"
+      "assembly. Generally, each input module would have been derived by compiling\n"
+      "a shader or pipeline with amdllpc, and using the -emit-lgc option to stop\n"
+      "before running LGC.\n"
+      "\n"
+      "The options section below lists all options available, including all LLVM ones.\n"
+      "The most relevant ones for the lgc utility are:\n"
+      "\n"
+      "  -extract=<index>         - Extract a single input module from multi-module\n"
+      "                             input. Index is 1-based\n"
+      "  -o=<filename>            - Output filename ('-' for stdout); defaults to the input\n"
+      "                             filename with the directory removed and the extension\n"
+      "                             replaced to match the output type\n"
+      "  -mcpu=<cpu-name>         - Target a specific GPU e.g. 'gfx900'\n"
+      "  -filetype=<value>        - Choose an output file type:\n"
+      "   =asm                    -   Emit an assembly ('.s') file (default)\n"
+      "   =obj                    -   Emit an ELF ('.elf') file\n"
+      "  -emit-llvm               - Emit LLVM IR assembly before running back-end\n"
+      "  -pal-abi-version=<uint>  - PAL pipeline ABI version to compile for (default\n"
+      "                             latest known)\n"
+      "  -verify-ir               - Verify IR after each pass\n"
+      );
+
```

Comment:
Would it make sense to enable `llvm::PrettyStackTraceProgram` here too? I expect the lgc tool to be used mostly for debugging/development, so increasing debuggability seems desirable to me.

Reply:
Interesting, I never felt that that option helps debuggability at all. All it does is make llvm-lit run *a lot* slower when processing many files with crashes... not that it really matters either way.

Reply:
I found it useful when browsing logs from other machines (think CI) where I cannot quickly rerun a failing test and attach to it.